### ISSUE TITLE
Machine "Provisioned" phase is never set and used.

### ISSUE THREAD

**Issue by onesolpark** [kubernetes-sigs/cluster-api#4779]
**What steps did you take and what happened:**
The machine object has the following phases to indicate the status of the lifecycle.
```Pending, Provisioning, Provisioned, Running, Deleting, Deleted, Failed, Unknown```

But it looks like machine can never be on the ```Provisioned``` phase.

https://github.com/kubernetes-sigs/cluster-api/blob/e970639fd4cb56003039878e82ffbc0b2e950a9e/controllers/machine_controller_phases.go#L61-L69

https://github.com/kubernetes-sigs/cluster-api/blob/e970639fd4cb56003039878e82ffbc0b2e950a9e/controllers/machine_controller_phases.go#L279-L311

From the codes above 
1) Machine is on ```Provisioned``` state if "nodeRef is set and infrastructure is not ready".
2) nodeRef is set only when infrastructure is in ```Ready``` state. 

In conclusion there is no case where the machine stays in ```Provisioned``` phase.

I am not sure if this is intended or a bug.

I am guessing ```m.Status.InfrastructureReady && m.Status.NodeRef == nil```
might be a case where machine is in the provisioned state.

**What did you expect to happen:**
Machine goes through the ```Provisioned``` phase.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version: master
- Kubernetes version: (use `kubectl version`): 1.21.0
- OS (e.g. from `/etc/os-release`): does not matter

/kind bug
/area machine

**Comment by vincepri** [comment]
> I am guessing m.Status.InfrastructureReady && m.Status.NodeRef == nil
might be a case where machine is in the provisioned state.

This is correct and the original reason for the "Provisioned" phase. The phase is meant to signify that the infrastructure was provisioned, but we haven't received yet a ProviderID from the infrastructure provider, which might be coming in later (for example bare metal providers).

**Comment by onesolpark** [comment]
Makes sense, but currently there is no way to reach the "Provisioned" phase.

As mentioned above, for ```m.Status.NodeRef != nil```  -> infrastructure needs to be ready.
Which will automatically turn the machine into "Running" phase without ever reaching into "Provisioned" phase.

Do you think changing ```m.Status.NodeRef != nil``` to ```m.Status.InfrastructureReady && m.Status.NodeRef == nil``` makes sense?

If so I will create a PR with the changes :)

**Comment by vincepri** [comment]
Going by the current definition in the proposal https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20190610-machine-states-preboot-bootstrapping.md#provisioned seems we should check the `ProviderID` to be present, NodeRef should come afterwards 🤔
